File: src/part2/ParticleFilter.jl
"""
    ParticleFilter(y, ye, param)

Particle filter for the project.
"""
function ParticleFilter(y, ye, param)
    w     = param.w                   # Parameter of the size of the FOV.
    P     = convert(Int64, param.P)   # Number of fish.
    N     = convert(Int64, param.N)   # Number of time snapshots.
    Np    = convert(Int64, param.Np)  # Number of particles per animal.
    ts    = param.ts                  # Time-step [s].
    σ_obs = param.sigma_obs           # Sd of the observation noise
                                      # on fish and enemy trajectories.
    k     = param.k                   # Shape parameter.
    s     = param.s                   # Scale parameter.

    d_y   = 1  # Dimension of the output space; must be 1 here.
    μ_w   = 0  # Mean of the gaussian error.
    t_f   = N * ts;

    # Probability density function of w_t, evaluated at w.
    α = 1. /√((2π)^d_y * abs(det(σ_obs)))
    out_noise_pdf(w) = α * exp(-0.5(w - μ_w)' * inv(σ_obs) * (w - μ_w))

    # Preallocate storage for the particles.
    X  = zeros(P, 2, Np, N+1)  # Fish particles.
    Xe = zeros(1, 2, Np, N+1)  # Predator particles.

    X̃  = zeros(P, 2, Np, N+1)  # Fish particles -- predictions.
    X̃e = zeros(1, 2, Np, N+1)  # Predator particles -- predictions.

    # Initialize orientations.
    # We assume that the orientation points
    # from the first observation to the second.
    o = y[:, :, 2] - y[:, :, 1]              # Generate random orientations.
    norm_o = .√(o[:, 1].^2 + o[:, 2].^2)     # Compute the norm.
    o[:, 1] = o[:, 1] ./ norm_o              # Normalize.
    o[:, 2] = o[:, 2] ./ norm_o

    oe = ye[:, :, 2] - ye[:, :, 1]           # Generate random orientations.
    norm_oe = .√(oe[:, 1].^2 + oe[:, 2].^2)  # Compute the norm.
    oe[:, 1] = oe[:, 1] ./ norm_oe           # Normalize.
    oe[:, 2] = oe[:, 2] ./ norm_oe

    # Initial sample set is the first observation for each particle.
    t = 0
    for i ∈ 1:Np
        for fish ∈ 1:P
            X[fish, :, i, t+1] = y[fish, :, i]
        end
        Xe[1, :, i, t+1] = ye[1, :, i]
    end

    # Start loop over time.
    println(N-1)
    for t ∈ 0:N-1
    println(t)
        # Prediction step.
        for i ∈ 1:Np
            # Use StateUpdate to get the next position/orientation
            # for the fish and the predator.
            mat"""
            [$next_x, $next_o, $next_xe, $next_oe] = StateUpdate($X(:, :, $i, $t+1), $o, $Xe(1, :, $i, $t+1), $oe, $ts, $k, $s, $w);
            """

            # Next fish orientations and positions.
            o = next_o
            X̃[:, :, i, t+1+1] = next_x

            # Next predator orientations and positions.
            oe = next_oe;
            X̃e[1, :, i, t+1+1] = next_xe
        end

        # Correction step.
        weights  = zeros(P, 2, Np)
        weightse = zeros(P, 2, Np)
        for i ∈ 1:Np
            for fish ∈ 1:P
                # Update weights for each fish.
                weights[fish, 1, i] = out_noise_pdf(y[fish, 1, t+1] - X̃[fish, 1, i, t+1+1])
                weights[fish, 2, i] = out_noise_pdf(y[fish, 2, t+1] - X̃[fish, 2, i, t+1+1])
            end
            # Update weights for the predator.
            weightse[1, 1, i] = out_noise_pdf(ye[1, 1, t+1] - X̃e[1, 1, i, t+1+1])
            weightse[1, 2, i] = out_noise_pdf(ye[1, 2, t+1] - X̃e[1, 2, i, t+1+1])
        end

        ind_sample  = zeros(Np, 2, P)
        ind_samplee = zeros(Np, 2, 1)
        for fish ∈ 1:P
            # Sample for the fish.
            ind_sample[:, 1, fish] = sample(collect(1:Np), pweights(weights[fish, 1, :]), Np)
            ind_sample[:, 2, fish] = sample(collect(1:Np), pweights(weights[fish, 2, :]), Np)
        end
        # Sample for the predator.
        ind_samplee[:, 1, 1] = sample(collect(1:Np), pweights(weightse[1, 1, :]), Np)
        ind_samplee[:, 2, 1] = sample(collect(1:Np), pweights(weightse[1, 2, :]), Np)

        for i ∈ 1:Np
            for fish ∈ 1:P
                # Update real positions with filtered positions for fish.
                X[fish, 1, i, t+1+1] = X̃[fish, 1, convert(Int64, ind_sample[i]), t+1+1]
                X[fish, 2, i, t+1+1] = X̃[fish, 2, convert(Int64, ind_sample[i]), t+1+1]
            end
            # Update real positions with filtered positions for predator.
            Xe[1, 1, i, t+1+1] = X̃e[1, 1, convert(Int64, ind_samplee[i]), t+1+1]
            Xe[1, 2, i, t+1+1] = X̃e[1, 2, convert(Int64, ind_samplee[i]), t+1+1]
        end
    end

    x̂  = zeros(P, 2, N)
    x̂e = zeros(1, 2, N)
    for i ∈ 1:N
        for fish ∈ 1:P
            # Update estimator for the fish.
            x̂[fish, 1, i] = mean(X[fish, 1, :, i])
            x̂[fish, 2, i] = mean(X[fish, 2, :, i])
        end
        # Update estimator for the predator.
        x̂e[1, 1, i] = mean(Xe[1, 1, :, i])
        x̂e[1, 2, i] = mean(Xe[1, 2, :, i])
    end
    return x̂, x̂e
end

File: src/part2/ParticleFilter.m
function [x_est,xe_est]= ParticleFilter(y,ye,param)

    w = param.w;         % Parameter of the size of the FOV.
    P = param.P;         % Number of fish.
    N = param.N;         % Number of time snapshots.
    Np = param.Np;       % Number of particles per animal.
    ts = param.ts;       % Time-step [s].
    sigma_obs = param.sigma_obs; % Sd of the observation noise
                                 % on fish and enemy trajectories.
    k = param.k;         % Shape parameter.
    s = param.s;         % Scale parameter
    d_y = 1;  % Dimension of output space; must be 1 in this script.
    mu_w = 0;
    t_f = N*ts;

    out_noise_pdf = @(w) 1/sqrt((2*pi)^d_y*abs(det(sigma_obs))) * exp(-.5*(w-mu_w)'*inv(sigma_obs)*(w-mu_w));  % pdf of the output noise w_t

    % Particles will be stored in X(e).
    X  = zeros(P,2,Np,N +1);
    Xe = zeros(1,2,Np,N +1);
    % Predictions are stored in Xtilde(e).
    Xtilde  = zeros(P,2,Np,N +1);
    Xtildee = zeros(1,2,Np,N +1);
    % Orientation for each particle
    o  = zeros(P,2,Np);
    oe = zeros(1,2,Np);

    % The initial orientation is computed as the difference between the first two observations, to which we add some noise.
    % This is a heuristic we found works pretty well.
    for i = 1:Np
        o(:,:,i) = y(:, :, 2) - y(:, :, 1) + normrnd(mu_w, sigma_obs, P, 2);
        norm_o = sqrt(o(:,1).^2 + o(:,2).^2);
        o(:,1,i) = o(:,1) ./ norm_o;
        o(:,2,i) = o(:,2) ./ norm_o;
        oe(1,:,i) = ye(:, :, 2) - ye(:, :, 1) + normrnd(mu_w, sigma_obs, 1, 2);
        norm_oe = sqrt(oe(:,1).^2 + oe(:,2).^2);
        oe(1,1,i) = oe(:,1) ./ norm_oe;
        oe(1,2,i) = oe(:,2) ./ norm_oe;
    end

    % The initial position is guessed by adding a random noise to the observations.
    t = 0;
    for i = 1:Np
        for fish = 1:P
            X(fish,:,i,t +1) = y(fish,:,1) + normrnd(mu_w, sigma_obs, 1, 2);
        end
        Xe(1,:,i,t +1) = ye(1,:,1) + normrnd(mu_w, sigma_obs, 1, 2);
    end


    % Start loop on time; N-2 because of array indexing starting at 1.
    for t = 0:N-2
        % Prediction step.
        % In order to improve performance, we compute the means,
        % and run the simulations with the meaned values.
        for i = 1:Np
        % Compute the next state for
        [next_x,next_o,next_xe,next_oe] = StateUpdate(X(:,:,i,t +1),o(:,:,i),Xe(1,:,i,t +1),oe(:,:,i),ts,k,s,w);
        o(:,:,i) = next_o;
        Xtilde(:,:,i,t+1 +1) = next_x;

        oe(:,:,i) = next_oe;
        Xtildee(1,:,i, t+1 +1) = next_xe;
        end

        % Correction step.
        % Compute the weights for resampling.
        weights  = zeros(P,2,Np);
        weightse = zeros(1,2,Np);
        for i=1:Np
            for fish = 1:P
                weights(fish,1,i) = out_noise_pdf(y(fish,1,t+1 +1)-Xtilde(fish,1,i,t+1 +1));
                weights(fish,2,i) = out_noise_pdf(y(fish,2,t+1 +1)-Xtilde(fish,2,i,t+1 +1));
            end
            weightse(1,1,i) = out_noise_pdf(ye(1,1,t+1 +1)-Xtildee(1,1,i,t+1 +1));
            weightse(1,2,i) = out_noise_pdf(ye(1,2,t+1 +1)-Xtildee(1,2,i,t+1 +1));
        end

        % Sampling.
        % We must take into account the possibility of rounding error on the sample weights leading to a zero vector.
        ind_sample  = zeros(P,Np,2);
        ind_samplee = zeros(1,Np,2);
        for fish = 1:P
            if sum(weights(fish, 1, :)) == 0.0
                ind_sample(fish,:,1) = randsample(Np,Np,true,ones(Np, 1));
            else
                ind_sample(fish,:,1) = randsample(Np,Np,true,weights(fish,1,:));
            end
            if sum(weights(fish, 2, :)) == 0.0
                ind_sample(fish,:,2) = randsample(Np,Np,true,ones(Np, 1));
            else
                ind_sample(fish,:,2) = randsample(Np,Np,true,weights(fish,2,:));
            end
        end
        if sum(weightse(1, 1, :)) == 0.0
            ind_samplee(1,:,1) = randsample(Np,Np,true,ones(Np, 1));
        else
            ind_samplee(1,:,1) = randsample(Np,Np,true,weightse(1,1,:));
        end
        if sum(weightse(1, 2, :)) == 0.0
            ind_samplee(1,:,2) = randsample(Np,Np,true,ones(Np, 1));
        else
            ind_samplee(1,:,2) = randsample(Np,Np,true,weightse(1,2,:));
        end

        % The estimated resampled position is stored in the position vector.
        for i=1:Np
            for fish = 1:P
                X(fish,1,i,t+1 +1) = Xtilde(fish,1,ind_sample(fish,i,1),t+1 +1);
                X(fish,2,i,t+1 +1) = Xtilde(fish,2,ind_sample(fish,i,2),t+1 +1);
            end
            Xe(1,1,i,t+1 +1) = Xtildee(1,1,ind_samplee(1,i,1),t+1 +1);
            Xe(1,2,i,t+1 +1) = Xtildee(1,2,ind_samplee(1,i,2),t+1 +1);
        end
    end

    % Return the mean of the resampled positions as the real estimate.
    x_est = zeros(P,2,N);
    xe_est = zeros(1,2,N);
    for i = 1:N
        for fish = 1:P
            x_est(fish,1,i) = mean(X(fish,1,:,i));
            x_est(fish,2,i) = mean(X(fish,2,:,i));
        end
        xe_est(1,1,i) = mean(Xe(1,1,:,i));
        xe_est(1,2,i) = mean(Xe(1,2,:,i));
    end
end
